tests/test_websockets.py

```diff
@@ -208,22 +207,24 @@ async def asgi(receive, send):
 
 def test_websocket_concurrency_pattern():
     def app(scope):
-        async def reader(websocket, queue):
-            async for data in websocket.iter_json():
-                await queue.put(data)
+        stream_send, stream_receive = anyio.create_memory_object_stream()
 
-        async def writer(websocket, queue):
-            while True:
-                message = await queue.get()
-                await websocket.send_json(message)
+        async def reader(websocket):
+            async with stream_send:
+                async for data in websocket.iter_json():
+                    await stream_send.send(data)
+
+        async def writer(websocket):
+            async with stream_receive:
+                async for message in stream_receive:
+                    await websocket.send_json(message)
 
         async def asgi(receive, send):
             websocket = WebSocket(scope, receive=receive, send=send)
-            queue = asyncio.Queue()
             await websocket.accept()
             await run_until_first_complete(
-                (reader, {"websocket": websocket, "queue": queue}),
-                (writer, {"websocket": websocket, "queue": queue}),
+                (reader, {"websocket": websocket}),
+                (writer, {"websocket": websocket}),
```

Comment:
I've not looked into it in detail, but I'd guess we can structure this more along structured concurrency lines, without my silly `run_until_first_complete` nonsense.

Eg. Using the recommended style in anyio... https://anyio.readthedocs.io/en/stable/streams.html?highlight=create_memory_object_stream#memory-object-streams

I'm also curious what the equivalent of `create_memory_object_stream` is in Trio.

Reply:
https://trio.readthedocs.io/en/stable/reference-core.html#trio.open_memory_channel

Reply:
@tomchristie I was able to simplify this further:
```python
        async def asgi(receive, send):
            websocket = WebSocket(scope, receive=receive, send=send)
            await websocket.accept()
            async with anyio.create_task_group() as task_group:
                task_group.start_soon(reader, websocket)
                task_group.start_soon(writer, websocket)
```

Reply:
You can save a task with:

```python
        async def asgi(receive, send):
            websocket = WebSocket(scope, receive=receive, send=send)
            await websocket.accept()
            async with anyio.create_task_group() as task_group:
                task_group.start_soon(reader, websocket)
                await writer(websocket)
```

Reply:
Fixed in dde5079bdf59c1f974dff30d888ba1062faf9145